// esbuild-node-externals/dist/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.findDependencies = exports.findPackagePaths = exports.createAllowPredicate = void 0;
const tslib_1 = require("tslib");
const path_1 = tslib_1.__importDefault(require("path"));
const fs_1 = tslib_1.__importDefault(require("fs"));
const find_up_1 = tslib_1.__importDefault(require("find-up"));
const createAllowPredicate = (allowList) => {
    return typeof allowList === 'function'
        ? allowList
        : (path) => Boolean(allowList.find((pattern) => typeof pattern === 'string' ? path === pattern : pattern.test(path)));
};
exports.createAllowPredicate = createAllowPredicate;
const isInDirectory = (parent, child) => {
    const relativePath = path_1.default.relative(parent, child);
    return !relativePath.startsWith('..') && !path_1.default.isAbsolute(relativePath);
};
const isInGitDirectory = (path, gitRootPath) => {
    return gitRootPath === undefined || isInDirectory(gitRootPath, path);
};
const findPackagePaths = (_cwd = process.cwd()) => {
    const gitDirectoryPath = find_up_1.default.sync('.git', {
        type: 'directory',
        cwd: _cwd,
    });
    const gitRootPath = gitDirectoryPath === undefined ? undefined : path_1.default.dirname(gitDirectoryPath);
    let cwd = _cwd;
    let packagePath;
    const packagePaths = [];
    while ((packagePath = find_up_1.default.sync('package.json', { type: 'file', cwd })) &&
        isInGitDirectory(packagePath, gitRootPath)) {
        packagePaths.push(packagePath);
        cwd = path_1.default.dirname(path_1.default.dirname(packagePath));
    }
    return packagePaths;
};
exports.findPackagePaths = findPackagePaths;
function getDependencyKeys(map = {}, allowWorkspaces = false) {
    if (!map) {
        return [];
    }
    if (!allowWorkspaces) {
        return Object.keys(map);
    }
    return Object.keys(map).filter((depKey) => !map[depKey].startsWith('workspace:'));
}
const findDependencies = (options) => {
    const packageJsonKeys = [
        options.dependencies && 'dependencies',
        options.devDependencies && 'devDependencies',
        options.peerDependencies && 'peerDependencies',
        options.optionalDependencies && 'optionalDependencies',
    ].filter(Boolean);
    const data = options.packagePaths.map((packagePath) => {
        let packageJson;
        try {
            const packageJsonString = fs_1.default.readFileSync(packagePath, 'utf8');
            packageJson = JSON.parse(packageJsonString);
        }
        catch (error) {
            console.error(error);
            throw new Error(`Couldn't process ${packagePath}". Make sure it's a valid JSON.`);
        }
        const packageNames = packageJsonKeys
            .map((key) => getDependencyKeys(packageJson[key], options.allowWorkspaces))
            .flat(1);
        const { allowPredicate } = options;
        return allowPredicate
            ? packageNames.filter((packageName) => !allowPredicate(packageName))
            : packageNames;
    });
    return data.flat(1);
};
exports.findDependencies = findDependencies;
